[query] StreamConstraintsException with large IRs in the ServiceBackend

### What happened?

From [zulip](https://hail.zulipchat.com/#narrow/channel/123010-Hail-Query-0.2E2-support/topic/StreamConstraintsException.20issue.20persisting.20in.200.2E2.2E133/near/480572744).

When initializing a QoB job, we parse the input json (as it contains the `rpcConfig`) before initializing a `ServiceBackend`, however as of #14651 we are setting this limit in the `Backend` constructor itself. This represents a regression from #14567.

https://github.com/hail-is/hail/blob/3fa74f091fe0eb8c0f0a24c2a0bad124f5d54da6/hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala#L458-L472

The solution is probably to restore this override in `ServiceBackendAPI.main`, or alternatively, refactor so that we create the `ServiceBackend` instance earlier.

### Version

0.2.133

### Relevant log output

```shell
2024-11-05 02:43:37.202 JVMEntryway: INFO: is.hail.JVMEntryway received arguments:
2024-11-05 02:43:37.202 JVMEntryway: INFO: 0: /hail-jars/gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar
2024-11-05 02:43:37.202 JVMEntryway: INFO: 1: is.hail.backend.service.Main
2024-11-05 02:43:37.202 JVMEntryway: INFO: 2: /batch/990e17d5209d429196c84ce010acab9d
2024-11-05 02:43:37.202 JVMEntryway: INFO: 3: /batch/990e17d5209d429196c84ce010acab9d/log
2024-11-05 02:43:37.202 JVMEntryway: INFO: 4: gs://hail-query-daaf463550/jars/4c60fddb171a52c21f41a81995c53a28e375c26b.jar
2024-11-05 02:43:37.202 JVMEntryway: INFO: 5: driver
2024-11-05 02:43:37.202 JVMEntryway: INFO: 6: execute(...)
2024-11-05 02:43:37.202 JVMEntryway: INFO: 7: gs://cpg-bioheart-hail/batch-tmp/tmp/hail/sRjJqvkZ3l9nmKuUErfNZv/jHpWQ6lemx/in
2024-11-05 02:43:37.202 JVMEntryway: INFO: 8: gs://cpg-bioheart-hail/batch-tmp/tmp/hail/sRjJqvkZ3l9nmKuUErfNZv/jHpWQ6lemx/out
2024-11-05 02:43:37.202 JVMEntryway: INFO: Yielding control to the QoB Job.
2024-11-05 02:43:37.206 ServiceBackendAPI$: INFO: BatchClient allocated.
2024-11-05 02:43:37.207 ServiceBackendAPI$: INFO: BatchConfig parsed.
2024-11-05 02:43:37.209 GoogleStorageFS$: INFO: Initializing google storage client from service account key
2024-11-05 02:43:37.783 JVMEntryway: ERROR: QoB Job threw an exception.
java.lang.reflect.InvocationTargetException: null
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at is.hail.JVMEntryway$1.run(JVMEntryway.java:119) [jvm-entryway.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: com.fasterxml.jackson.core.exc.StreamConstraintsException: String length (20013488) exceeds the maximum length (20000000)
	at com.fasterxml.jackson.core.StreamReadConstraints.validateStringLength(StreamReadConstraints.java:324) ~[jackson-core-2.15.2.jar:2.15.2]
	at com.fasterxml.jackson.core.util.ReadConstrainedTextBuffer.validateStringLength(ReadConstrainedTextBuffer.java:27) ~[jackson-core-2.15.2.jar:2.15.2]
	at com.fasterxml.jackson.core.util.TextBuffer.finishCurrentSegment(TextBuffer.java:939) ~[jackson-core-2.15.2.jar:2.15.2]
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._finishString2(UTF8StreamJsonParser.java:2584) ~[jackson-core-2.15.2.jar:2.15.2]
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._finishAndReturnString(UTF8StreamJsonParser.java:2560) ~[jackson-core-2.15.2.jar:2.15.2]
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.getText(UTF8StreamJsonParser.java:335) ~[jackson-core-2.15.2.jar:2.15.2]
	at is.hail.relocated.org.json4s.jackson.JValueDeserializer._deserialize$1(JValueDeserializer.scala:26) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.relocated.org.json4s.jackson.JValueDeserializer._deserialize$1(JValueDeserializer.scala:48) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.relocated.org.json4s.jackson.JValueDeserializer._deserialize$1(JValueDeserializer.scala:48) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.relocated.org.json4s.jackson.JValueDeserializer._deserialize$1(JValueDeserializer.scala:48) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.relocated.org.json4s.jackson.JValueDeserializer.deserialize(JValueDeserializer.scala:57) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323) ~[jackson-databind-2.15.2.jar:2.15.2]
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:2105) ~[jackson-databind-2.15.2.jar:2.15.2]
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1481) ~[jackson-databind-2.15.2.jar:2.15.2]
	at is.hail.relocated.org.json4s.jackson.JsonMethods.parse(JsonMethods.scala:35) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.relocated.org.json4s.jackson.JsonMethods.parse$(JsonMethods.scala:20) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.relocated.org.json4s.jackson.JsonMethods$.parse(JsonMethods.scala:71) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.backend.service.ServiceBackendAPI$.$anonfun$main$5(ServiceBackend.scala:467) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.utils.package$.using(package.scala:673) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.backend.service.ServiceBackendAPI$.main(ServiceBackend.scala:467) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.backend.service.Main$.main(Main.scala:10) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	at is.hail.backend.service.Main.main(Main.scala) ~[gs:__hail-query-daaf463550_jars_4c60fddb171a52c21f41a81995c53a28e375c26b.jar.jar:?]
	... 12 more
2024-11-05 02:43:37.787 JVMEntryway: ERROR: Exception encountered in QoB cancel thread.
org.newsclub.net.unix.SocketClosedException: Not open
	at org.newsclub.net.unix.AFCore.validFdOrException(AFCore.java:90) ~[jvm-entryway.jar:?]
	at org.newsclub.net.unix.AFSocketImpl$AFInputStreamImpl.read(AFSocketImpl.java:510) ~[jvm-entryway.jar:?]
	at java.io.DataInputStream.readInt(DataInputStream.java:393) ~[?:?]
	at is.hail.JVMEntryway$2.run(JVMEntryway.java:136) [jvm-entryway.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
```

--- a/hail/python/test/hail/test_ir.py
+++ b/hail/python/test/hail/test_ir.py
@@ -1,4 +1,5 @@
 import functools
+import random
 import re
 import unittest
 from test.hail.helpers import resource, skip_unless_spark_backend, skip_when_service_backend
@@ -774,3 +775,12 @@ def test_locus_interval_encoding():
     _assert_encoding_roundtrip(start)
     _assert_encoding_roundtrip(end)
     _assert_encoding_roundtrip(interval)
+
+
+def test_very_large_ir_deserializes():
+    # see: https://github.com/hail-is/hail/issues/14580
+    #      https://github.com/hail-is/hail/issues/14650
+    large_list = [random.getrandbits(63) for _ in range(5_000_000)]
+    large_lit = hl.literal(large_list, hl.tarray(hl.tint64))
+    round_trip = hl.eval(large_lit)
+    assert large_list == round_trip

--- a/hail/src/main/scala/is/hail/backend/service/Main.scala
+++ b/hail/src/main/scala/is/hail/backend/service/Main.scala
@@ -1,13 +1,26 @@
 package is.hail.backend.service
 
+import com.fasterxml.jackson.core.StreamReadConstraints
+
 object Main {
   val WORKER = "worker"
   val DRIVER = "driver"
 
-  def main(argv: Array[String]): Unit =
+  /* This constraint should be overridden as early as possible. See:
+   * - https://github.com/hail-is/hail/issues/14580
+   * - https://github.com/hail-is/hail/issues/14749
+   * - The note on this setting in is.hail.backend.Backend
+   */
+  StreamReadConstraints.overrideDefaultStreamReadConstraints(
+    StreamReadConstraints.builder().maxStringLength(Integer.MAX_VALUE).build()
+  )
+
+  def main(argv: Array[String]): Unit = {
+
     argv(3) match {
       case WORKER => Worker.main(argv)
       case DRIVER => ServiceBackendAPI.main(argv)
       case kind => throw new RuntimeException(s"unknown kind: $kind")
     }
+  }
 }

--- a/hail/src/main/scala/is/hail/backend/service/Main.scala
+++ b/hail/src/main/scala/is/hail/backend/service/Main.scala
@@ -9,18 +9,15 @@ object Main {
   /* This constraint should be overridden as early as possible. See:
    * - https://github.com/hail-is/hail/issues/14580
    * - https://github.com/hail-is/hail/issues/14749
-   * - The note on this setting in is.hail.backend.Backend
-   */
+   * - The note on this setting in is.hail.backend.Backend */
   StreamReadConstraints.overrideDefaultStreamReadConstraints(
     StreamReadConstraints.builder().maxStringLength(Integer.MAX_VALUE).build()
   )
 
-  def main(argv: Array[String]): Unit = {
-
+  def main(argv: Array[String]): Unit =
     argv(3) match {
       case WORKER => Worker.main(argv)
       case DRIVER => ServiceBackendAPI.main(argv)
       case kind => throw new RuntimeException(s"unknown kind: $kind")
     }
-  }
 }

